Diagnostic for .unwrap_or_default() suggests removing unrelated code

### Code

```Rust
fn main() {
    let mut tx_heights: std::collections::BTreeMap<(), Option<()>> = <_>::default();
    tx_heights.get(&()).unwrap_or_default();
}
```

### Current output

```Shell
error[E0277]: the trait bound `&Option<()>: Default` is not satisfied
 --> a.rs:3:25
  |
3 |     tx_heights.get(&()).unwrap_or_default();
  |                         ^^^^^^^^^^^^^^^^^ the trait `Default` is not implemented for `&Option<()>`
  |
  = help: the trait `Default` is implemented for `Option<T>`
note: required by a bound in `Option::<T>::unwrap_or_default`
 --> /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/core/src/option.rs:1073:5
help: consider removing this method call, as the receiver has type `BTreeMap<(), Option<()>>` and `BTreeMap<(), Option<()>>: Default` trivially holds
  |
3 -     tx_heights.get(&()).unwrap_or_default();
3 +     tx_heights.unwrap_or_default();
  |

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0277`.
```
nightly's even worse:
```Shell
error[E0277]: the trait bound `&Option<()>: Default` is not satisfied
    --> a.rs:3:25
     |
   3 |     tx_heights.get(&()).unwrap_or_default();
     |                         ^^^^^^^^^^^^^^^^^ the trait `Default` is not implemented for `&Option<()>`
     |
help: the trait `Default` is implemented for `Option<T>`
    --> /mnt/filling/store/nabijaczleweli/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:2226:1
     |
2226 | impl<T> const Default for Option<T> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: required by a bound in `Option::<T>::unwrap_or_default`
    --> /mnt/filling/store/nabijaczleweli/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:1095:12
     |
1093 |     pub const fn unwrap_or_default(self) -> T
     |                  ----------------- required by a bound in this associated function
1094 |     where
1095 |         T: [const] Default,
     |            ^^^^^^^^^^^^^^^ required by this bound in `Option::<T>::unwrap_or_default`
help: consider removing this method call, as the receiver has type `BTreeMap<(), Option<()>>` and `BTreeMap<(), Option<()>>: Default` trivially holds
     |
   3 -     tx_heights.get(&()).unwrap_or_default();
   3 +     tx_heights.unwrap_or_default();
     |

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0277`.
```

### Desired output

Not this.

### Rust Version

```Shell
rustc 1.93.0-nightly (cc3eee7fb 2025-11-28)
binary: rustc
commit-hash: cc3eee7fbe17ea4b7238531cb97e1b7b8bd6afce
commit-date: 2025-11-28
host: x86_64-unknown-linux-gnu
release: 1.93.0-nightly
LLVM version: 21.1.5
```

### Anything else?

_No response_

--- a/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs
@@ -4069,24 +4069,45 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                 ))
                 && expr.span.hi() != rcvr.span.hi()
             {
-                match tcx.hir_node(call_hir_id) {
-                    // Do not suggest removing a method call if the argument is the receiver of the parent call:
-                    // `x.a().b()`, suggesting removing `.a()` would change the type and could make `.b()` unavailable.
+                let should_sugg = match tcx.hir_node(call_hir_id) {
                     Node::Expr(hir::Expr {
                         kind: hir::ExprKind::MethodCall(_, call_receiver, _, _),
                         ..
-                    }) if call_receiver.hir_id == arg_hir_id => {}
-                    _ => {
-                        err.span_suggestion_verbose(
-                            expr.span.with_lo(rcvr.span.hi()),
-                            format!(
-                                "consider removing this method call, as the receiver has type `{ty}` and \
-                                `{pred}` trivially holds",
-                            ),
-                            "",
-                            Applicability::MaybeIncorrect,
-                        );
+                    }) if let Some((DefKind::AssocFn, did)) =
+                        typeck_results.type_dependent_def(call_hir_id)
+                        && call_receiver.hir_id == arg_hir_id =>
+                    {
+                        // Avoid suggesting removing a method call if the argument is the receiver of the parent call and
+                        // removing the receiver would make the method inaccessible. i.e. `x.a().b()`, suggesting removing
+                        // `.a()` could change the type and make `.b()` unavailable.
+                        if tcx.inherent_impl_of_assoc(did).is_some() {
+                            // if we're calling an inherent impl method, just try to make sure that the receiver type stays the same.
+                            Some(ty) == typeck_results.node_type_opt(arg_hir_id)
+                        } else {
+                            // we're calling a trait method, so we just check removing the method call still satisfies the trait.
+                            let trait_id = tcx
+                                .trait_of_assoc(did)
+                                .unwrap_or_else(|| tcx.impl_trait_id(tcx.parent(did)));
+                            let args = typeck_results.node_args(call_hir_id);
+                            let tr = ty::TraitRef::from_assoc(tcx, trait_id, args)
+                                .with_replaced_self_ty(tcx, ty);
+                            self.type_implements_trait(tr.def_id, tr.args, param_env)
+                                .must_apply_modulo_regions()
+                        }
                     }
+                    _ => true,
+                };
+
+                if should_sugg {
+                    err.span_suggestion_verbose(
+                        expr.span.with_lo(rcvr.span.hi()),
+                        format!(
+                            "consider removing this method call, as the receiver has type `{ty}` and \
+                            `{pred}` trivially holds",
+                        ),
+                        "",
+                        Applicability::MaybeIncorrect,
+                    );
                 }
             }
             if let hir::Expr { kind: hir::ExprKind::Block(block, _), .. } = expr {

--- a/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs
@@ -4069,15 +4069,25 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                 ))
                 && expr.span.hi() != rcvr.span.hi()
             {
-                err.span_suggestion_verbose(
-                    expr.span.with_lo(rcvr.span.hi()),
-                    format!(
-                        "consider removing this method call, as the receiver has type `{ty}` and \
-                         `{pred}` trivially holds",
-                    ),
-                    "",
-                    Applicability::MaybeIncorrect,
-                );
+                match tcx.hir_node(call_hir_id) {
+                    // Do not suggest removing a method call if the argument is the receiver of the parent call:
+                    // `x.a().b()`, suggesting removing `.a()` would change the type and could make `.b()` unavailable.
+                    Node::Expr(hir::Expr {
+                        kind: hir::ExprKind::MethodCall(_, call_receiver, _, _),
+                        ..
+                    }) if call_receiver.hir_id == arg_hir_id => {}
+                    _ => {
+                        err.span_suggestion_verbose(
+                            expr.span.with_lo(rcvr.span.hi()),
+                            format!(
+                                "consider removing this method call, as the receiver has type `{ty}` and \
+                                `{pred}` trivially holds",
+                            ),
+                            "",
+                            Applicability::MaybeIncorrect,
+                        );
+                    }
+                }
             }
             if let hir::Expr { kind: hir::ExprKind::Block(block, _), .. } = expr {
                 let inner_expr = expr.peel_blocks();

--- a/tests/ui/trait-bounds/argument-with-unnecessary-method-call.rs
+++ b/tests/ui/trait-bounds/argument-with-unnecessary-method-call.rs
@@ -9,3 +9,11 @@ fn main() {
     //~| HELP try using a fully qualified path to specify the expected types
     //~| HELP consider removing this method call, as the receiver has type `Bar` and `Bar: From<Bar>` trivially holds
 }
+
+// regression test for https://github.com/rust-lang/rust/issues/149487.
+fn quux() {
+    let mut tx_heights: std::collections::BTreeMap<(), Option<()>> = <_>::default();
+    tx_heights.get(&()).unwrap_or_default();
+    //~^ ERROR the trait bound `&Option<()>: Default` is not satisfied
+    //~| HELP: the trait `Default` is implemented for `Option<T>`
+}

--- a/tests/ui/trait-bounds/argument-with-unnecessary-method-call.stderr
+++ b/tests/ui/trait-bounds/argument-with-unnecessary-method-call.stderr
@@ -23,6 +23,18 @@ LL -     qux(Bar.into());
 LL +     qux(Bar);
    |
 
-error: aborting due to 1 previous error
+error[E0277]: the trait bound `&Option<()>: Default` is not satisfied
+  --> $DIR/argument-with-unnecessary-method-call.rs:16:25
+   |
+LL |     tx_heights.get(&()).unwrap_or_default();
+   |                         ^^^^^^^^^^^^^^^^^ the trait `Default` is not implemented for `&Option<()>`
+   |
+help: the trait `Default` is implemented for `Option<T>`
+  --> $SRC_DIR/core/src/option.rs:LL:COL
+note: required by a bound in `Option::<T>::unwrap_or_default`
+  --> $SRC_DIR/core/src/option.rs:LL:COL
 
-For more information about this error, try `rustc --explain E0283`.
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0277, E0283.
+For more information about an error, try `rustc --explain E0277`.

